### PR TITLE
rename EigenTensor to math::Tensor

### PR DESCRIPTION
**Pull request by QiJune**
Since [EigenVector](https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors) is actually a mathematical concepts.  It's better to find a more appropriate name. Here, just make a new namespace paddle::framework::math to hold Eigen related types.

### CODE REVIEW

File: paddle/framework/tensor.h [hunk 1]
```diff
@@ -32,16 +32,23 @@ struct CastToPyBufferImpl;
...
   template <bool less, size_t i, typename... args>
   friend struct paddle::pybind::details::CastToPyBufferImpl;
 
   template <typename T, size_t D, int MajorType, typename IndexType>
-  friend struct EigenTensor;
+  friend struct paddle::framework::math::Tensor;
 
   template <typename T, int MajorType, typename IndexType>
-  friend struct EigenVector;
+  friend struct paddle::framework::math::Vector;
```
Comment (jacquesqiao) [inline paddle/framework/tensor.h]: if there is a way to put `friend struct paddle::framework::math::Matrix` here？

Comment (jacquesqiao) [inline paddle/framework/tensor.h]: change [following code](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/eigen.h#L82)
```cpp
template <typename T, int MajorType = Eigen::RowMajor,
          typename IndexType = Eigen::DenseIndex>
using EigenMatrix = EigenTensor<T, 2, MajorType, IndexType>;
```
to
```cpp
template <typename T, int MajorType = Eigen::RowMajor,
          typename IndexType = Eigen::DenseIndex>
struct EigenMatrix : EigenTensor<T, 2, MajorType, IndexType> {};
```
and add
```cpp
  template <typename T, int MajorType, typename IndexType>
  friend struct EigenMatrix;
```
to tensor.h will make the code more consistency

**Comment by wangkuiyi** [comment]
Eigen is more specific and accurate as a name than `math`. I don't think we need the rename.

**Review by wangkuiyi — COMMENTED** [review]
Eigen is more specific and accurate as a name than math. I don't think we need the rename.

**Comment by helinwang** [comment]
> Eigen is more specific and accurate as a name than math. I don't think we need the rename.

@wangkuiyi I do feel that `EigenVector` as a class name is a little confusing, since "eigen vector" is a linear algebra concept, closely related to vector and matrix calculation.
If `eigen` is more specific and accurate than `math`, how do you think about: `framework::eigen::vector`, or `framework::eigen3::vector`?

**Pull request closed without merging**